concourse/client/util.py

```diff
@@ -122,3 +137,118 @@ def jobs_not_triggered(
+                concourse_api=concourse_api,
+            ):
+                concourse_api.trigger_build(
+                    pipeline_name=resource.pipeline_name(),
+                    job_name=job.pipeline_name,
+                )
+
+                if wait_for_job_to_be_triggered(
+                    job=job,
+                    resource_version=resource_version,
+                    concourse_api=concourse_api,
+                    retries=10,
+                    sleep_time_seconds=5
+                ):
+                    info(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        'has been triggered'
+                    )
+                else:
+                    warning(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        'could not be triggered'
+                    )
+                return
+        except PinningFailed:
+            retries -= 1
+
+    warning(f'Could not process untriggered job {job.name}')
+
+
+@contextmanager
+def pin_resource(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    pin_comment = {
+        'datetime': datetime.now().isoformat(),
+        'resource_type': resource.type().value,
+        'version_data': resource_version.version(),
+        'comment': 'automatically pinned by technical user, please do not unpin manually',
+    }
+    _ensure_resource_unpinned(
+        resource=resource,
+        concourse_api=concourse_api,
+        retries=10,
+        sleep_time_seconds=5,
+    )
+    try:
+        concourse_api.pin_and_comment_resource_version(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+            resource_version_id=resource_version,
+            comment=json.dumps(pin_comment),
+        )
+        yield
+    finally:
+        concourse_api.unpin_resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+
+def _ensure_resource_unpinned(
+    resource: PipelineResource,
+    concourse_api,
+    retries: int=10,
+    sleep_time_seconds: int=5,
+):
+    while retries >= 0:
+        cc_resource = concourse_api.resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+        if not cc_resource.is_pinned():
+            return
+
+        pin_comment = cc_resource.pin_comment()
+        try:
+            # if comment is json format, we assume it was pinned by us
+            parsed_comment = json.loads(pin_comment)
+            pin_time = datetime.fromisoformat(parsed_comment['datetime'])
+            if pin_time + timedelta(minutes=3) < datetime.now():
+                # resource pinned by us but unpin did not happen for more than 3 minutes -> cleanup
+                concourse_api.unpin_resource(
```

Comment:
this should definitely be logged as a warning, as it indicates a bug (previous job did not free its "lock"). also, there should be a coupling of maximum retries and maximum waiting time. currently, this is just hardcoded at two places, which I find to be far from ideal.

-> consider writing remaining retries into pinning comment. is it possible to update a pinning comment w/o releasing the pin? if so, please update each time the triggering was retried, and update the "last-tried" timestamp (+ store the sleep-time). in this case, we do not even have to hardcode anything here.

instead, we can check:

- last time of update
- sleep-time

if last-time-of-update + sleep-time > now, then something went wrong.